File: bench/bench_ocaml/test/test_float.ml
module Float = Math_lib.Float
module Test = Math_lib.Common.Test

let _test_scientific () =
  let a : Float.t = Float.inverse (Float.new' (Z.of_int 6) Z.zero) in
  let value_resolved = Float.resolve a (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int 166) in

  let a : Float.t = Float.inverse (Float.new' (Z.of_int 6) Z.zero) in
  let a : Float.t = Float.mul (Float.new' (Z.of_int (-1)) Z.zero) a in
  let value_resolved = Float.resolve a (Z.of_int 3) in
  let () = assert (value_resolved = Z.of_int (-166)) in

  let a : Float.t = Float.inverse (Float.new' (Z.of_int 3) Z.zero) in
  let b : Float.t = Float.inverse (Float.new' (Z.of_int 2) Z.zero) in
  let value : Float.t = Float.add a b in
  (* //let () = assert(value = { p=5; q=6 }) in *)
  let value_resolved = Float.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = (Z.of_int 833)) in

  let a : Float.t = Float.inverse (Float.new' (Z.of_int 3) Z.zero) in
  let b : Float.t = Float.inverse (Float.new' (Z.of_int 2) Z.zero) in
  let value : Float.t = Float.sub a b in
  (* //let () = assert(value = { p=-1; q=6 }) in *)
  let value_resolved = Float.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = (Z.of_int (-166))) in

  let a : Float.t = Float.inverse (Float.new' (Z.of_int 3) Z.zero) in
  let b : Float.t = Float.inverse (Float.new' (Z.of_int 2) Z.zero) in
  let value : Float.t = Float.mul a b in
  (* //let () = assert(value = { p=1; q=6 }) in *)
  let value_resolved = Float.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = (Z.of_int 166)) in

  let a : Float.t = Float.inverse (Float.new' (Z.of_int 3) Z.zero) in
  let b : Float.t = Float.inverse (Float.new' (Z.of_int 2) Z.zero) in
  let value : Float.t = Float.div a b in
  (* //let () = assert(value = { p=2; q=3 }) in *)
  let value_resolved = Float.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = (Z.of_int 666)) in

  (* // 1/2 % 1/3 *)
  let a : Float.t = Float.inverse (Float.new' (Z.of_int 2) Z.zero) in
  let b : Float.t = Float.inverse (Float.new' (Z.of_int 3) Z.zero) in
  let value : Float.t = Float.modulo a b in
  (* //let () = assert(value = { p=2; q=3 }) in *)
  let value_resolved = Float.resolve value (Z.of_int 3) in
  let () = assert (value_resolved = (Z.of_int 166)) in

  Test.log "Test finished"

let _ = _test_scientific ()

File: bench/bench_kotlin/src/test/kotlin/math/TestFloat.kt
package math

import math.TestFloat.Float.add
import math.TestFloat.Float.inverse
import math.TestFloat.Float.mul
import math.TestFloat.Float.new
import math.TestFloat.Float.sub
import org.openjdk.jmh.annotations.Benchmark
import org.openjdk.jmh.annotations.Scope
import org.openjdk.jmh.annotations.State
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals

@State(Scope.Benchmark)
open class TestFloat {

    object Float : math.Float

    @Test
    fun testScientific00() {
        val a = inverse(new(6))
        val resolved = Float.resolve(a)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(166))
    }

    @Test
    fun testScientific01() {
        val a = inverse(new(6))
        val b = mul(new(-1))(a)
        val resolved = Float.resolve(b)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(-166))
    }

    @Test
    fun testScientific02() {
        val a = inverse(new(3))
        val b = inverse(new(2))
        val v = add(a)(b)
        val resolved = Float.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(833))
    }

    @Test
    fun testScientific03() {
        val a = inverse(new(3))
        val b = inverse(new(2))
        val v = sub(a)(b)
        val resolved = Float.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(-166))
    }

    @Test
    fun testScientific04() {
        val a = inverse(new(3))
        val b = inverse(new(2))
        val v = mul(a)(b)
        val resolved = Float.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(166))
    }

    @Test
    fun testScientific05() {
        val a = inverse(new(3))
        val b = inverse(new(2))
        val v = Float.div(a)(b)
        val resolved = Float.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(666)) // Hell yeah
    }

    @Test
    fun testScientific06() {
        val a = inverse(new(2))
        val b = inverse(new(3))
        val v = Float.modulo(a)(b)
        val resolved = Float.resolve(v)(BigInteger.valueOf(3))
        assertEquals(resolved, BigInteger.valueOf(166)) // Hell yeah
    }

    @Benchmark
    fun all() {
        testScientific00()
        testScientific01()
        testScientific02()
        testScientific03()
        testScientific04()
        testScientific05()
        testScientific06()
    }
}